core/src/test/java/org/apache/calcite/test/enumerable/EnumerableHashJoinTest.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.test.enumerable;
+
+import org.apache.calcite.adapter.java.ReflectiveSchema;
+import org.apache.calcite.config.CalciteConnectionProperty;
+import org.apache.calcite.config.Lex;
+import org.apache.calcite.test.CalciteAssert;
+import org.apache.calcite.test.JdbcTest;
+
+import org.junit.Test;
+
+/**
+ * Unit test for
+ * {@link org.apache.calcite.adapter.enumerable.EnumerableHashJoin}.
+ */
+public class EnumerableHashJoinTest {
```

Comment:
Could we also add here some tests for SEMI/ANTI join with/without predicate?

Reply:
ok, will do it

Reply:
Sorry, I just saw there are already the ANTI join tests in place:
`EnumerableJoinTest#equiAntiJoin`, `EnumerableJoinTest#nonEquiAntiJoin`, so I would say no need to add new tests for the ANTI join case. For the SEMI equi-nonEqui you could just add new tests in that class.

Reply:
Please see EnumerableHashJoinTest , but I can't make an assert to explain the plan  when useing RelBuilder to build a semi join.

Reply:
Yes, as far as I know, this is a limitation (can't verify the plan if you use `withRel`). In any case, the returned result is validated, so I guess that should be fine.